### PR TITLE
Evicting unused cache entries before RasterCache actions

### PR DESCRIPTION
**Pull request by ColdPaleLight**
Delete the unused images before generating new cached images.
fix: https://github.com/flutter/flutter/issues/94511

1. Add new method 'RasterCache::EvictUnusedCacheEntries', called before the LayerTree::TryToRasterCache method to remove cached images that are no longer used.

2. Tweak `total_count` and `total_bytes` of the `RasterCacheMetrics`.
``` diff
- size_t total_count() const { return in_use_count + eviction_count; }
+ size_t total_count() const { return in_use_count; }

- size_t total_bytes() const { return in_use_bytes + eviction_bytes; }
+ size_t total_bytes() const { return in_use_bytes; }
```

3. Rename some methods.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by ColdPaleLight** [comment]
cc @flar @JsouLiang

**Review by flar — CHANGES_REQUESTED** [review]
This was one of my most anticipated follow-on opts for the refactor. It's looking good. Some minor nits that I'd like to hear back about before I approve.

**Comment by JsouLiang** [comment]
Could we provide some benchmarks data?

File: flow/testing/mock_raster_cache.cc
Comment (ColdPaleLight) [reply flow/testing/mock_raster_cache.cc:29]: `AddMockLayer` and `AddMockPicture` in `MockRasterCache` are used to force the layer or display list to be cached in the raster cache. The reason why `MarkSeen` needs to be called here is that layer does not meet the following conditions.
```
  if (num_cache_attempts_ >= layer_cached_threshold_) {
    // the layer can be cached
    cache_state_ = CacheState::kCurrent;
    context->raster_cache->MarkSeen(key_id_, matrix_, true);
  } 
```
The current implementation of method `AddMockPicture` is also the same, manually calling `MarkSeen`.

File: flow/raster_cache.cc
Comment (ColdPaleLight) [reply flow/raster_cache.cc:171]: Done. I added the lifecycle documentation to `raster_cache.h`. Since my English is limited, I would appreciate if you could help improve it.

**Comment by flar** [comment]
Hey, as an FYI - this utility provided in the framework repo can be used to format the A/B results in a more readable format:

https://github.com/flutter/flutter/blob/7148cba6bbe113eecc8b7fdce9b970cdc83edeb8/dev/devicelab/bin/summarize.dart

File: flow/raster_cache.h
Comment (flar) [inline flow/raster_cache.h:99]: Fix the indentation here and on the line below. Otherwise, great comment!

File: flow/raster_cache.h
Comment (flar) [inline flow/raster_cache.h:102]: Fix indentation here too.

File: flow/raster_cache.h
Comment (flar) [inline flow/raster_cache.h:97]: Perhaps indent this one more level and indicate the recursion:

- Preroll stage
  - LayerTree::Preroll - for each Layer in the tree:
    - RCI::Setup ...
    - RCI::Finalize ...

File: flow/raster_cache.h
Comment (flar) [inline flow/raster_cache.h:108]: - LayerTree::Paint - for each layer in the tree:

File: flow/testing/mock_raster_cache.cc
Comment (flar) [reply flow/testing/mock_raster_cache.cc:29]: The behavior in AddMockPicture is also a little "bypassing the mechanisms". Really the loop in AddMockPicture could be using Preroll to do the same.

The issue here is that we only preroll the layer once, but it has a threshold for caching the layer itself. Calling Preroll in a loop would also meet that threshold, wouldn't it?

So, both should be using a loop calling Preroll, with the thresholds used to control the loop.

Comment (flar) [reply flow/testing/mock_raster_cache.cc:29]: AddMockPicture also calls MarkSeen followed by Draw, but I don't think the call to Draw is necessary any more as I moved the increment of `access_count` (now known as `accesses_since_visible`) from Draw to MarkSeen as part of https://github.com/flutter/engine/pull/34562 and so you only need to Preroll now to meet the threshold.

**Review by flar — CHANGES_REQUESTED** [review]

File: flow/raster_cache.h
Comment (ColdPaleLight) [reply flow/raster_cache.h]: Done.

File: flow/raster_cache.h
Comment (ColdPaleLight) [reply flow/raster_cache.h]: Done.

File: flow/raster_cache.h
Comment (ColdPaleLight) [reply flow/raster_cache.h]: Done.

File: flow/raster_cache.h
Comment (ColdPaleLight) [reply flow/raster_cache.h]: Done.

Comment (ColdPaleLight) [reply flow/raster_cache.h]: Done.

File: flow/testing/mock_raster_cache.cc
Comment (ColdPaleLight) [reply flow/testing/mock_raster_cache.cc]: Done. Remove `MarkSeen` calls form both `AddMockPicture` and `AddMockLayer`.  `AddMockLayer` ends up not using the loop, I set `layer_cached_threshold` to 1 for layer instead.

**Review by flar — APPROVED** [review]
LGTM

**Comment by flar** [comment]
I added the autosubmit label to the benchmark to get it into the working tree ASAP to collect baseline data, but I'll leave it to you to tag this PR at your discretion. During the time that this takes to just roll into the main tree we'll likely get a few samples of baseline, but it would be nice to have a couple dozen so that the regression measuring code has some statistically relevant data to compare against.

**Comment by flar** [comment]
Benchmark results look good here: https://flutter-flutter-perf.skia.org/e/?begin=1658253277&end=1658571306&queries=test%3Dcolor_filter_with_unstable_child_perf__e2e_summary&requestType=0

**Comment by zanderso** [comment]
More results from SkiaPerf: https://flutter-flutter-perf.skia.org/e/?begin=1658344386&end=1658770538&keys=Xbdb54ced983e1377840e8781177ff1b2&requestType=0&xbaroffset=30039

**Pull request merged**
